errext/exit_code.go

```diff
@@ -0,0 +1,66 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package errext
+
+import "errors"
+
+// ExitCode is the code with which the application should exit if this error
+// bubbles up to the top of the scope. Values should be between 0 and 125:
+// https://unix.stackexchange.com/questions/418784/what-is-the-min-and-max-values-of-exit-codes-in-linux
+type ExitCode uint8
+
+// HasExitCode is a wrapper around an error with an attached exit code.
+type HasExitCode interface {
+	error
+	ExitCode() ExitCode
+}
+
+// WithExitCode can attach an exit code to the given error, if it doesn't have
+// one already. It won't do anything if the error already had an exit code
+// attached. Similarly, if there is no error (i.e. the given error is nil), it
+// also won't do anything.
+func WithExitCode(err error, exitCode ExitCode) error {
+	if err == nil {
+		// No error, do nothing
+		return nil
+	}
+	var ecerr HasExitCode
+	if errors.As(err, &ecerr) {
+		// The given error already has an exit code, do nothing
+		return err
+	}
```

Comment:
I think I would never expect this to happen.

Is there a situation where that is a problem?

Reply:
I use this property in https://github.com/k6io/k6/blob/1d741a854b1b18cd3af99f397d857375a630648c/cmd/run.go#L248-L252

The idea is that, since you're wrapping errors, you generally want to keep the most specific exit code possible. So having the `WithExitCode` not overwrite an already existing attached error code is very helpful - in the above code, if we already had a an exit code (e.g. a script error) we'll exit with its code (e.g. `107`), but if we don't, we'd still attach a the more generic one there.

Reply:
And, of course, you can always negate that with a custom error type (or another helper), if some other use case requires masking of a previously attached exit code. But this makes sense to me as the default behavior.

Reply:
I am totally not sold on the "This is already wrapped in error fo the "same type" so I will silently not wrap it for you" idea. IMO w/e code will exit with the code should check whether it was wrapped multiple times and decide whether it should exit with the most inner one or not. Or for the code that is wrapping to check if it has ErrorCode and either wrap it again ... or not.

I am win with `WithExitCodeIfNotOneAlready` being added

Reply:
hmm fair enough... I am not sure if I should just rename this to `WithExitCodeIfNotOneAlready` / `WithExitCodeIfNone`, since we don't really have a use case for something that always attaches an exit code yet, in most places it either doesn't matter (root errors deep in the code that first attach an exit code) or we don't want to overwrite it

alternatively, I can rip it out and only check for existing exit codes in a few places, I'll see how that looks like

Reply:
Is this ok? https://github.com/k6io/k6/pull/2046/commits/d8e41ec4a5b7394a192dc5410ebf5e647cab242d